File: extractor2/src/extractor.cpp
#include "extractor.hpp"

#include <iostream>
#include <fstream>
#include <filesystem>

#include "utils.hpp"
#include "json.hpp"
using json = nlohmann::json;

const char* PGM_VERSION = "0.0.1";
const char* PGM_DESCRIPTION = ""
"\n"
"Description:\n"
"  Ylands Export Extractor v2\n"
"  Used with Ylands editor tools:\n"
"    - \"EXPORTBLOCKDEFS.ytool\"\n"
"    - \"EXPORTSCENE.ytool\"\n"
"  to pull JSON data from log file \"log_userscript_ct.txt\".\n"
"  Accepts existing JSON exports for direct conversion.\n"
"  Depending on options can output:\n"
"    - Clean JSON ready for custom scene reconstruction or later conversion.\n"
"    - 3D model (GLB, GLTF, or OBJ).\n"
"  Configure with accompanying file \"extract_config.json\" and \"lookup.json\".\n"
"  See README for more details.\n"
"  Authors: BinarySemaphore\n"
"  Updated: 2025-03-04\n";
const char* PGM_OPTIONS_HELP = ""
"\n"
"Options:\n"
"          -v, --version : Show info.\n"
"             -h, --help : Show info and options help.\n"
"-i, --input <JSON-file> : (optional) Read from existing JSON.\n"
"                          Takes existing JSON export from Ylands.\n"
"                          If not given, will attempt to extract from Ylands\n"
"                          directly using settings in \"extract_config.json\".\n"
"    -o, --output <name> : (optional) Output name [no file extension]\n"
"                          (defualt: \"output\").\n"
"                          Output files will automatically have appropriate\n"
"                          extensions based on TYPE.\n"
"      -t, --type <TYPE> : (optional) Output type\n"
"                          (default: JSON).\n"
"                          Supported TYPEs : file(s) created.\n"
"                             GLB : <name>.glb\n"
"                            GLTF : <name>.gltf and <name>.bin\n"
"                             OBJ : <name>.obj and <name>.mtl\n"
"                            JSON : <name>.json\n"
"                          See \"Output Options\" for more additional options.\n"
"                          If input data is Block Definitions from\n"
"                          \"EXPORTBLOCKDEFS.ytool\" output TYPE forced as JSON.\n"
"\n"
"Output Options (only applies to TYPEs: GLB, GLTF, and JSON):\n"
"  -m : Merge into single geometry.\n"
"       Same as using '-rja'\n"
"  -r : Remove internal faces within same material (unless using -a).\n"
"       Any faces adjacent and opposite another face are removed.\n"
"       This includes their opposing neighbor's face.\n"
"  -j : Join verticies within same material (unless using -a)\n"
"       Any vertices sharing a location with another, or within a very small\n"
"       distance, will be reduced to a single vertex.\n"
"       This efectively \"hardens\" or \"joins\" Yland blocks into a single\n"
"       geometry.\n"
"  -a : Apply to all.\n"
"       For any Join (-j) or Internal Face Removal (-r).\n"
"       Applies to all regardless of material grouping.\n";
/*
Hidden Options (for post-build):
--preload <file> : Run YlandStandard::preload_lookups(<file>).
				   Iterates through refs, loading *.obj files.
				   Substitues the lookup with loaded objects.
				   Serializes the result into a binary file
				   to be loaded by enduser at runtime.
*/
const char* CONFIG_FILE = "./extract_config.json";
const char* LOG_STRIP = "# ";
const char* DATA_INDICATOR_START = "# }";
const char* DATA_INDICATOR_BLOCKDEF = "Export Block Ref (below):";
const char* DATA_INDICATOR_SCENE = "Export Scene (below):";
const char* DATA_TYPE_BLOCKDEF = "BLOCKDEF";
const char* DATA_TYPE_SCENE = "SCENE";
const char* CF_KEY_INSTALL_DIR = "Ylands Install Location";
const char* CF_KEY_LOG_PATH = "Log Location";
const char* CF_KEY_AUTO_NEST = "Auto Nest Scenes";
const char* CF_KEY_PPRINT = "Output Pretty";

void printHelp(const char* pgm_name) {
	printHelp(pgm_name, false);
}

void printHelp(const char* pgm_name, bool version_only) {
	std::cout << "Program: " << pgm_name << std::endl << std::endl;
	std::cout << "Version: " << PGM_VERSION << std::endl;
	std::cout << PGM_DESCRIPTION;
	if (!version_only) {
		std::cout << PGM_OPTIONS_HELP;
	}
}

Config getConfigFromArgs(int argc, char** argv) {
	bool get_preload_filename = false;
	bool get_input_filename = false;
	bool get_output_filename = false;
	bool get_export_type = false;
	bool missing_arg = false;
	char missing_arg_name[25];
	char missing_arg_expects[100];
	char pgm_name[250];
	Config config;

	// Defaults
	config.output_filename = "output";
	config.export_type = ExportType::JSON;
	config.preload = false;
	config.has_input = false;
	config.remove_faces = false;
	config.join_verts = false;
	config.apply_all = false;

	for (int i = 0; i < argc; i++) {
		if (i == 0) {
			std::strcpy(pgm_name, argv[0]);
			continue;
		}

		// Get secondary args
		if (get_preload_filename) {
			config.preload_filename = argv[i];
			get_preload_filename = false;
		} else if (get_input_filename) {
			config.input_filename = argv[i];
			get_input_filename = false;
		} else if (get_output_filename) {
			config.output_filename = argv[i];
			get_output_filename = false;
		} else if (get_export_type) {
			if (std::strcmp(argv[i], "GLB") == 0 || std::strcmp(argv[i], "glb") == 0) {
				config.export_type = ExportType::GLB;
				get_export_type = false;
			} else if (std::strcmp(argv[i], "GLTF") == 0 || std::strcmp(argv[i], "gltf") == 0) {
				config.export_type = ExportType::GLTF;
				get_export_type = false;
			} else if (std::strcmp(argv[i], "OBJ") == 0 || std::strcmp(argv[i], "obj") == 0) {
				config.export_type = ExportType::OBJ;
				get_export_type = false;
			} else if (std::strcmp(argv[i], "JSON") == 0 || std::strcmp(argv[i], "json") == 0) {
				config.export_type = ExportType::JSON;
				get_export_type = false;
			}
		}

		// Catch primary args
		else if (std::strcmp(argv[i], "--help") == 0 || std::strcmp(argv[i], "-h") == 0) {
			printHelp(pgm_name);
			exit(0);
		} else if (std::strcmp(argv[i], "--version") == 0 || std::strcmp(argv[i], "-v") == 0) {
			printHelp(pgm_name, true);
			exit(0);
		} else if (std::strcmp(argv[i], "--preload") == 0) {
			config.preload = true;
			get_preload_filename = true;
		} else if (std::strcmp(argv[i], "--input") == 0 || std::strcmp(argv[i], "-i") == 0) {
			config.has_input = true;
			get_input_filename = true;
		} else if (std::strcmp(argv[i], "--output") == 0 || std::strcmp(argv[i], "-o") == 0) {
			get_output_filename = true;
		} else if (std::strcmp(argv[i], "--type") == 0 || std::strcmp(argv[i], "-t") == 0) {
			get_export_type = true;
		} else if (std::strcmp(argv[i], "-r") == 0) {
			config.remove_faces = true;
		} else if (std::strcmp(argv[i], "-j") == 0) {
			config.join_verts = true;
		} else if (std::strcmp(argv[i], "-a") == 0) {
			config.apply_all = true;
		} else if (std::strcmp(argv[i], "-m") == 0) {
			config.remove_faces = true;
			config.join_verts = true;
			config.apply_all = true;
		} else {
			std::cerr << "Unrecongnized argument \"" << argv[i] << "\". "
					  << "Use -h or --help to show options." << std::endl;
			exit(1);
		}
	}

	// Check if secondary args missing
	missing_arg = get_preload_filename || get_input_filename ||
				  get_output_filename || get_export_type;
	if (get_preload_filename) {
		std::strcpy(missing_arg_name, "--preload");
		std::strcpy(missing_arg_expects, "a filename");
	} else if (get_input_filename) {
		std::strcpy(missing_arg_name, "--input (-i)");
		std::strcpy(missing_arg_expects, "a filename");
	} else if (get_output_filename) {
		std::strcpy(missing_arg_name, "--output (-o)");
		std::strcpy(missing_arg_expects, "a filename");
	} else if (get_export_type) {
		std::strcpy(missing_arg_name, "--type (-t)");
		std::strcpy(missing_arg_expects, "a Type (JSON, GLB, GLTF, or OBJ)");
	}
	if (missing_arg) {
		std::cerr << "Arguement missing: \"" << missing_arg_name
				  << "\" expects " << missing_arg_expects << std::endl;
		exit(1);
	}

	return config;
}

int extractAndExport(Config& config) {
	json data;

	if (!config.has_input) {
		try {
			extractFromYlands(config, data);
		} catch (CustomException& e) {
			std::cerr << "Error while extrating from Ylands: " << e.what()
					  << std::endl;
			return 1;
		}
		if (isDataBlockDef(data) && config.export_type != ExportType::JSON) {
			std::cout << "Extracted " << DATA_TYPE_BLOCKDEF
					  << " from Ylands, forcing JSON export." << std::endl;
			config.export_type = ExportType::JSON;
		}
	} else {
		try {
			loadFromFile(config.input_filename.c_str(), data);
		} catch (CustomException& e) {
			std::cerr << "Error while loading from JSON file \""
					  << config.input_filename << "\": " << e.what()
					  << std::endl;
			return 1;
		}
		if (config.export_type == ExportType::JSON) {
			std::cout << "Nothing to do:"
					  << " Loading from JSON file and exporting to JSON."
					  << std::endl;
			return 0;
		}
		if (isDataBlockDef(data)) {
			std::cerr << "Error: Given Ylands BLOCKDEF data." << std::endl
					  << "BLOCKDEF is for referencing only, not processing."
					  << std::endl
					  << "To set BLOCKDEF update \"lookup.json\" file."
					  << std::endl;
			return 2;
		}
	}

	// JSON export
	if (config.export_type == ExportType::JSON) {
		try {
			exportAsJson(config.output_filename.c_str(), data, config.ext_pprint);
		} catch(CustomException& e) {
			std::cerr << "Error exporting JSON file \""
					  << config.output_filename << "\": "
					  << e.what() << std::endl;
			return 3;
		}
		return 0;
	}

	if (!isDataValidScene(data)) {
		std::cerr << "Invalid JSON data: expected Ylands SCENE data."
				  << std::endl;
		return 2;
	}

	// Convert data into 3D Scene
	// Process scene using config flags

	// OBJ export
	// GLTF export
	// GLB export

	// Finish
	return 0;
}

void exportAsJson(const char* filename, const json& data, bool pprint) {
	char filename_ext[200] = "";
	std::strcat(filename_ext, filename);
	std::strcat(filename_ext, ".json");

	std::cout << "Exporting [JSON] file \"" << filename_ext << "\"..." << std::endl;
	std::ofstream f(filename_ext);
	if (!f.is_open()) {
		throw SaveException("Failed to open file");
	}
	if (pprint) {
		f << data.dump(4);
	} else {
		f << data.dump();
	}
	f.close();
	std::cout << "Export complete" << std::endl << std::endl;
}

void loadFromFile(const char* filename, json& data) {
	std::cout << "Loading from JSON file \"" << filename << "\"..." << std::endl;
	std::ifstream f(filename);
	if (!f.is_open()) {
		throw LoadException("Failed to open file");
	}
	try {
		data = json::parse(f);
	} catch (json::parse_error& e) {
		f.close();
		throw ParseException("Failed to parse JSON file: " + std::string(e.what()));
	}
	f.close();
	std::cout << "Loaded from JSON file" << std::endl << std::endl;
}

void extractFromYlands(Config& config, json& data) {
	bool data_type_set;
	int i;
	int data_start = -1;
	char confirm[2];
	char data_type[20] = "";
	std::filesystem::path ylands_install_dir;
	std::filesystem::path log_fullpath;
	std::string line;
	std::string raw_data;
	std::vector<std::string> log_lines;
	std::vector<std::string> raw_data_lines;

	std::cout << "Extracting directly from Ylands..." << std::endl;
	updateConfigFromFile(config, CONFIG_FILE);

	ylands_install_dir = std::filesystem::path(config.ylands_install_dir);
	log_fullpath = std::filesystem::path(config.ylands_log_path);
	log_fullpath = ylands_install_dir / log_fullpath;

	std::cout << "Loading log file \"" << log_fullpath.string() << "\"..." << std::endl;
	std::ifstream f(log_fullpath);
	if (!f.is_open()) {
		throw LoadException("Failed to open \"" + log_fullpath.string() + "\".");
	}
	while (std::getline(f, line)) {
		log_lines.push_back(line);
	}
	if (log_lines.size() == 0) {
		throw ParseException("Log file is empty");
	}
	f.close();
	std::cout << "Loaded" << std::endl << std::endl;

	std::cout << "Searching for exported data in log file..." << std::endl;
	data_type_set = false;
	for (i = log_lines.size() - 1; i >= 0; i--) {
		line = log_lines[i];
		if (line.rfind(DATA_INDICATOR_START, 0) == 0) {
			data_start = i;
		}
		if (data_start >= 0 && !data_type_set) {
			if (line.find(DATA_INDICATOR_BLOCKDEF) != std::string::npos) {
				std::strcpy(data_type, DATA_TYPE_BLOCKDEF);
				data_type_set = true;
			} else if (line.find(DATA_INDICATOR_SCENE) != std::string::npos) {
				std::strcpy(data_type, DATA_TYPE_SCENE);
				data_type_set = true;
			}
		}
		if (data_type_set) {
			raw_data_lines = std::vector<std::string>(
				log_lines.begin() + i + 1,
				log_lines.begin() + data_start + 1
			);
			log_lines.clear();
			break;
		}
	}
	if (!data_type_set || raw_data_lines.size() == 0) {
		throw ParseException(
			"Could not find any exported data, please try restarting"
			" Ylands and rerunning the export tool"
		);
	}
	std::cout << "Found " << data_type << " exported data"
			  << std::endl << std::endl;
	
	std::cout << "Processing " << data_type << " data to JSON..." << std::endl;
	for (i = 0; i < raw_data_lines.size(); i++) {
		raw_data_lines[i] = string_replace(raw_data_lines[i], LOG_STRIP, "");
	}
	raw_data = string_join(raw_data_lines, "");
	try {
		data = json::parse(raw_data);
	} catch (json::parse_error& e) {
		std::cerr << "Failed to parse JSON data: " << e.what() << std::endl;
		std::cerr << "Storing raw data in \"error.txt\"" << std::endl;
		std::ofstream f("error.txt");
		if (!f.is_open()) {
			throw SaveException("Failed to open \"error.txt\" for writing");
		}
		f << raw_data;
		f.close();
		throw ParseException(
			"Failed to parse JSON data: see error/log for details"
		);
	}
	raw_data_lines.clear();
	raw_data.clear();
	std::cout << "JSON ready" << std::endl << std::endl;

	if (std::strcmp(data_type, DATA_TYPE_SCENE) == 0 && config.ext_auto_nest) {
		std::cout << "Reformat " << data_type << " JSON from flat to nested..."
				  << std::endl;
		try {
			reformatSceneFlatToNested(data);
		} catch (ParseException& e) {
			std::cerr << "Failed to reformat data: " << e.what() << std::endl;
			std::cout << "Continue without nesting? (y/n): ";
			std::cin >> confirm;
			std::cout << std::endl;
			if (confirm[0] != 'y') {
				throw ParseException(
					"Failed to reformat data (user prompted: abort and exit):"
					" see error/log for details"
				);
			}
		}
		std::cout << "JSON data reformatted" << std::endl << std::endl;
	}
	std::cout << "JSON extracted from Ylands" << std::endl << std::endl;
}

void reformatSceneFlatToNested(json& data) {
	std::string parent_key;
	std::vector<std::string> keys_to_remove;
	json* parent;

	for (auto& [key, item] : data.items()) {
		if (item.contains("parent")) {
			parent_key = item["parent"];
			if (!data.contains(parent_key)) {
				throw ParseException(
					"Cannot find parent \"" + parent_key + "\" in JSON data"
				);
			}
			parent = &data[parent_key];
			if (!parent->contains("children")) {
				(*parent)["children"] = json {};
			}
			(*parent)["children"][key] = item;
			keys_to_remove.push_back(key);
		}
	}

	for (int i = 0; i < keys_to_remove.size(); i++) {
		data.erase(keys_to_remove[i]);
	}
}

void updateConfigFromFile(Config& config, const char* filename) {
	json data;

	std::cout << "Loading config file: \"" << filename << "\"..." << std::endl;
	std::ifstream f(filename);
	if (!f.is_open()) {
		throw LoadException("Failed to open \"" + std::string(filename)
		+ "\".");
	}
	try {
		data = json::parse(f);
	} catch (json::parse_error& e) {
		f.close();
		throw ParseException("Failed to parse JSON file: " + std::string(e.what()));
	}
	f.close();
	if (data.contains(CF_KEY_INSTALL_DIR)) {
		config.ylands_install_dir = data[CF_KEY_INSTALL_DIR];
	} else config.ylands_install_dir = "";
	if (data.contains(CF_KEY_LOG_PATH)) {
		config.ylands_log_path = data[CF_KEY_LOG_PATH];
	} else config.ylands_log_path = "";
	if (data.contains(CF_KEY_AUTO_NEST)) {
		config.ext_auto_nest = data[CF_KEY_AUTO_NEST];
	} else config.ext_auto_nest = true;
	if (data.contains(CF_KEY_PPRINT)) {
		config.ext_pprint = data[CF_KEY_PPRINT];
	} else config.ext_pprint = false;

	validateConfigAndPromptForFixes(config, filename, false);

	// Update data with any changes to config from validating
	data[CF_KEY_INSTALL_DIR] = config.ylands_install_dir;
	data[CF_KEY_LOG_PATH] = config.ylands_log_path;
	data[CF_KEY_AUTO_NEST] = config.ext_auto_nest;
	data[CF_KEY_PPRINT] = config.ext_pprint;
	std::cout << "Config loaded: " << data.dump(4).c_str() << std::endl;
}

void validateConfigAndPromptForFixes(Config& config, const char* filename, bool config_changed) {
	bool valid = true;
	char confirm[2];

	std::filesystem::path install_dir(config.ylands_install_dir);
	std::filesystem::path log_path(config.ylands_log_path);
	std::filesystem::path log_fullpath = install_dir / log_path;

	if (!std::filesystem::exists(install_dir) ||
		!std::filesystem::is_directory(install_dir)) {
		std::cout << "\nInvalid Config \"" << CF_KEY_INSTALL_DIR
				  << "\", directory does not exist." << std::endl;
		std::cout << "Path was: \"" << config.ylands_install_dir
				  << "\"." << std::endl;
		std::cout << "Enter new path: ";
		getline(std::cin, config.ylands_install_dir);
		std::cout << std::endl;
		config_changed = true;
		valid = false;
	} else if (!std::filesystem::exists(log_fullpath) ||
		!std::filesystem::is_regular_file(log_fullpath)) {
		std::cout << "\nInvalid Config \"" << CF_KEY_LOG_PATH
				  << "\", file does not exist." << std::endl;
		std::cout << "Path was: \"" << config.ylands_log_path
				  << "\"." << std::endl;
		std::cout << "Enter new path: ";
		getline(std::cin, config.ylands_log_path);
		std::cout << std::endl;
		config_changed = true;
		valid = false;
	}

	if (!valid) {
		return validateConfigAndPromptForFixes(config, filename, config_changed);
	}

	if (config_changed) {
		std::cout << "Config was updated, save to config file? (y/n): ";
		std::cin >> confirm;
		std::cout << std::endl;
		if (confirm[0] == 'y' || confirm[0] == 'Y') {
			std::ofstream f(filename);
			if (!f.is_open()) {
				throw SaveException(
					"Failed to open \"" + std::string(filename)
					+ "\" for writing."
				);
			}
			f << "{\n"
			  << "    \"" << CF_KEY_INSTALL_DIR << "\": \""
			  << config.ylands_install_dir << "\",\n"
			  << "    \"" << CF_KEY_LOG_PATH << "\": \""
			  << config.ylands_log_path << "\",\n"
			  << "    \"" << CF_KEY_AUTO_NEST << "\": ";
			if (config.ext_auto_nest) f << "true";
			else f << "false";
			f << ",\n";
			f << "    \"" << CF_KEY_PPRINT << "\": ";
			if (config.ext_pprint) f << "true";
			else f << "false";
			f << "\n";
			f << "}";
			f.close();
			std::cout << "Updated config saved to \"" << filename
					  << "\"" << std::endl;
		}
	}
}

bool isDataBlockDef(const json& data) {
	bool is_blockdef = false;
	for (auto& [key, value] : data.items()) {
		if (value.contains("shape") || value.contains("NONE")) {
			is_blockdef = true;
		}
		break;
	}
	return is_blockdef;
}


bool isDataValidScene(const json& data) {
	bool is_validscene = false;
	for (auto& [key, value] : data.items()) {
		if (value.contains("position") || value.contains("rotation")) {
			is_validscene = true;
		}
		break;
	}
	return is_validscene;
}

File: core/extractor.py
#!python
'''
Ylands Export Extractor
Used with Ylands tools ("EXPORTBLOCKDEFS.ytool" and "EXPORTSCENE.ytool") to
pull JSON data from log file "log_userscript_ct.txt".
Produces clean JSON ready for scene reconstruction.
Configure with accompanying file "config.json".
See README for more details.
Author: BinarySemaphore
Updated: 2025-03-13
'''
__version__ = '0.1.2'

import os
import sys
import json

CONFIG_FILE = "./config.json"
LOG_STRIP = "# "
DATA_INDICATOR_START = "# }"
DATA_INDICATOR_BLOCKDEF = "Export Block Ref (below):"
DATA_INDICATOR_SCENE = "Export Scene (below):"
DATA_TYPE_BLOCKDEF = "BLOCKDEF"
DATA_TYPE_SCENE = "SCENE"

def extract(config, output_filename="out.json"):
    '''
    :param dict config: Required key-values:
                        'ylands_base_dir': "<ylands-install-dir-fullpath>"
                        'ylands_script_log': "<userscript-log-file-relpath>"
                        Optional key-values:
                        'nest_scene': <bool>
                        'pprint': <bool>
                        
    :param str output_filename: JSON filename to save extracted data into
    :returns: exist status
    :rtype: int
    '''
    cf_base_dir = config['ylands_base_dir']
    cf_log = config['ylands_script_log']
    cf_nest = config['nest_scene']
    cf_pprint = config['pprint']
    
    yland_log_lines = []
    yland_log_file = os.path.join(cf_base_dir, cf_log)
    data_start = -1
    data_raw_lines = []
    data_type = None
    
    print("Loading log file \"%s\"..." % yland_log_file)
    with open(yland_log_file, 'r', encoding="ascii", errors="replace") as f:
        yland_log_lines = f.readlines()
    if not yland_log_lines:
        print("No data from log file")
        return 1
    yland_log_lines.reverse()
    print("Loaded\n")
    
    print("Searching for exported data...")
    for index in range(len(yland_log_lines)):
        line = yland_log_lines[index]
        if line.startswith(DATA_INDICATOR_START):
            data_start = index
        if data_start >= 0 and not data_type:
            if DATA_INDICATOR_BLOCKDEF in line:
                data_type = DATA_TYPE_BLOCKDEF
            elif DATA_INDICATOR_SCENE in line:
                data_type = DATA_TYPE_SCENE
        if data_type:
            data_raw_lines = yland_log_lines[data_start:index]
            del yland_log_lines
            break
    if not data_type or not data_raw_lines:
        print("Could not find any export data, please try restarting Ylands and rerunning the export tool")
        return 1
    data_raw_lines.reverse()
    print("Found %s exported data\n" % data_type)
    
    print("Processing %s data to JSON..." % data_type)
    for index in range(len(data_raw_lines)):
        data_raw_lines[index] = data_raw_lines[index].strip(LOG_STRIP)
    data_raw = "".join(data_raw_lines)
    data_raw = data_raw.encode("ascii", errors="replace").decode("ascii")
    data = None
    try:
        data = json.loads(data_raw)
    except json.JSONDecodeError as e:
        print("Failed get JSON data: %s" % e)
        print("Storing raw data in \"error.txt\"")
        with open("error.txt", 'w') as f:
            f.write(data_raw)
        return 2
    del data_raw_lines
    del data_raw
    print("JSON ready\n")
    
    if data_type == "SCENE" and cf_nest:
        print("Reformat %s JSON from flat to nested..." % data_type)
        try:
            data = reformatSceneFlatToNested(data)
        except Exception as e:
            confirm = input("Failed to reformat data, continue without nesting? (y/n): ")
            if confirm != 'y':
                print("Abort and exit")
                return 3
        print("JSON data reformatted\n")
    
    print("Writing JSON to file \"%s\"..." % output_filename)
    while(os.path.exists(output_filename)):
        confirm = input("Output file already exists, overwrite? (y/n): ")
        if confirm == 'y':
            break
        output_filename = input("Rename output file (leave empty to quit): ")
        if not output_filename:
            print("Abort write and exit")
            return 0
    with open(output_filename, 'w') as f:
        if cf_pprint:
            json.dump(data, f, indent=4)
        else:
            json.dump(data, f)
    print("JSON for %s written to file\n" % data_type)
    
    print("Done")
    return 0

def reformatSceneFlatToNested(data):
    reformated_data = {}
    for key in data:
        item = data[key]
        if item.get('parent'):
            parent_key = item['parent']
            parent = data.get(parent_key)
            if not parent or not isinstance(parent, dict):
                raise Exception("Cannot find parent \"%s\" in source data" % parent_key)
            if not parent.get('children'):
                parent['children'] = {}
            parent['children'][key] = item
            del parent['children'][key]['parent']
        else:
            reformated_data[key] = item
    return reformated_data

def configFromFilePromptIfNeeded(config_file):
    print("Loading config file: \"%s\"..." % config_file)
    
    config_data = None
    with open(config_file, 'r') as f:
        config_data = json.load(f)
    
    config = loadAndValidateConfigDataOfferSave(config_file, config_data)
    
    print("Config loaded: %s\n" % json.dumps(config_data, indent=4))
    return config

def loadAndValidateConfigDataOfferSave(config_file, config_data):
    config = {}
    save_config = False
    base_dir = None
    # Map to internal config key and validation method
    config_mapping = {
        "Ylands Install Location": ("ylands_base_dir", os.path.isdir),
        "Log Location": ("ylands_script_log", os.path.exists),
        "Auto Nest Scenes": ("nest_scene", validateIgnore),
        "Output Pretty": ("pprint", validateIgnore)
    }
    
    for data_key in config_mapping:
        config_key, validate_method = config_mapping[data_key]
        config[config_key] = config_data.get(data_key)
        
        validate_value = config[config_key]
        if config_key == "ylands_base_dir":
            base_dir = config[config_key]
        elif config_key == "ylands_script_log" and validate_value:
            validate_value = os.path.join(base_dir, validate_value)
        
        if validate_value is None or not validate_method(validate_value):
            print("\nInvalid Config \"%s\": \"%s\"" % (data_key, config[config_key]))
            config[config_key] = input("Enter new value for \"%s\": " % data_key)
            save_config = True
    
    if save_config:
        confirm = input("Config was updated, save to config file? (y/n): ")
        if confirm == 'y':
            for data_key in config_mapping:
                config_key, _ignore = config_mapping[data_key]
                config_data[data_key] = config[config_key]
            with open(config_file, 'w') as f:
                json.dump(config_data, f, indent=4)
            print("Updated config saved to \"%s\"" % config_file)
        return loadAndValidateConfigDataOfferSave(config_file, config_data)
    else:
        return config

def validateIgnore(_ignore):
    return True

def argParse():
    config = {
        'output_filename': 'out.json'
    }
    
    if '-h' in sys.argv or '--help' in sys.argv:
        printHelp()
        exit(0)
    
    if '-o' in sys.argv:
        config['output_filename'] = str(sys.argv[sys.argv.index('-o')+1])
    
    return config

def printHelp():
    script_file = os.path.basename(__file__)
    print(__doc__)
    print("Version: %s" % __version__)
    print("\nUsage:")
    print("  python %s [options]" % script_file)
    print("  ./%s [options]" % script_file)
    print("\nArguments:")
    print("  NONE : Configuration in \"%s\"" % CONFIG_FILE)
    print("\nOptions:")
    print("         -h --help : Display this information/help.")
    print("  -o <output-file> : Specify output filename (default: out.json).")
    print("\nExamples:")
    print("1) ./%s" % script_file)
    print("2) python %s -o myhouse.json" % script_file)
    print("\nExample \"%s\"" % CONFIG_FILE)
    print('''{
    "Ylands Install Location": "C:\\Program Files (x86)\\Steam\\steamapps\\common\\Ylands",
    "Log Location": "Ylands_Data\\log_userscript_ct.txt",
    "Auto Nest Scenes": true,
    "Output Pretty": false
}''')

if __name__ == "__main__":
    config_term = argParse()
    config_data = configFromFilePromptIfNeeded(CONFIG_FILE)
    status = extract(config_data, config_term['output_filename'])
    exit(status)
